src/components/select/multi-select/multi-select.component.js

```diff
@@ -111,6 +116,34 @@ const MultiSelect = React.forwardRef(
       [name, id]
     );
 
+    /* generic value update function which can be used for both controlled and uncontrolled
+     * components, both with and without onChange.
+     * - if onChange is provided, calls that with the updated value, provided it has changed in length
+     * (ie if an option has been selected or deselected)
+     * - if the component is uncontrolled, call setSelectedValue with the provided update function
+     * - if the component is controlled, do not change selectedValue and instead directly call
+     * the update function and pass its result to onChange
+     * - ensure that in every circumstance, onChange and the update function are only called at most
+     * once, in case they have side effects */
+    const setValue = useCallback(
+      (updateFunction) => {
+        const updateAndCallOnChange = (oldValue) => {
+          const newValue = updateFunction(oldValue);
+          if (onChange && newValue.length !== oldValue.length) {
+            onChange(createCustomEvent(newValue));
+          }
+          return newValue;
+        };
+
+        if (isControlled.current) {
+          updateAndCallOnChange(value);
+        } else {
+          setSelectedValue(updateAndCallOnChange);
```

Comment:
Hi Kuba, sorry I only just got to this - setState functions in React accept a function as an argument in order to update the value according to that function. We were already doing this in MultiSelect - see line 419 of the old file for example - I just refactored into a common utility so that we could easily treat both controlled and uncontrolled cases correctly.